shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -621,6 +621,18 @@ - (void)insertText:(NSString*)text {
 
 - (void)deleteBackward {
   _selectionAffinity = _kTextAffinityDownstream;
+
+  // When deleting a combining character, _selectedTextRange has location
+  // but does not have length, so we have to manually set it.
+  if (_selectedTextRange.isEmpty && [self hasText]) {
+    NSRange oldRange = ((FlutterTextRange*)_selectedTextRange).range;
+    if (oldRange.location > 0) {
+      NSRange newRange = NSMakeRange(oldRange.location - 1, 1);
```

Comment:
This looks reasonable, but would it be better to use `(UITextRange*)characterRangeByExtendingPosition:inDirection:` here? That method is expressly designed to handle multi-byte character sequences. (via the `incrementPosition:` and `decrementPosition:` methods).

 I'm not sure what the expected behaviour is with Thai combining characters, but I'd be curious if those methods produce the correct result. They're generally a bit safer.

Reply:
I found that I couldn't use `(UITextRange*)characterRangeByExtendingPosition:inDirection:` here.

This method returns the length of grapheme cluster, however, in case of Thai, (and also Arabic), we wanted to delete only a part of grapheme cluster, which is totally opposite to multibyte emoji. 😭

Reply:
I have tested this, primarily manually, and verified this fixes the inability to delete when a vowel is typed bug. I did not notice any impact on other languages/emojis/multibyte sequences, although I have not comprehensively surveyed all the possible ways unicode can work.

In this case, removing one codepoint is indeed what we want, we just have to be careful this isn't overly broad and will regress other languages.

I will do some more testing/invesitigation to convince myself that this is (relatively) safe.

Reply:
@HelloCore thanks for verifying! Yeah that method is intended to create ranges that always go to the bounds of a grapheme cluster. My worry was that the patch as-is would cause problems for other cases (e.g. emoji with skin-tone etc.) but @GaryQian has confirmed it appears to not cause issues with anything he's tested.

Would you mind adding an explanatory comment for this code that includes a link back to the original issues (flutter/flutter#24203, flutter/flutter#21745) to help future authors/maintainers of this code to understand what problem it's solving.

I don't know enough about Thai to have a good understanding of how input works and what is expected on iOS and other platforms, but my intuition says there's probably an underlying issue in the way we've implemented mark regions on iOS. Since it looks like this patch resolves the issue and doesn't appear to cause any issues with other cases. I'm fine with us landing this (once the comments are added).

But I think we should file a bug to have someone spend a few days understanding how Thai input is expected to work, especially with regards to the mark region (iOS)/composing region (Android). In our testing on Android in particular, it looks wrong to me, but my only real experience with the composing region is implementing support for Japanese, which I've got much better understanding of.

Thanks for your contribution, and for your patience!

Reply:
FYI: I've filed https://github.com/flutter/flutter/issues/28962

Reply:
I've pushed an update to the comment.

Reply:
Thank you for reviewing.

As I investigated, I found that `UIKeyboardImpl` called  
- `- (UITextPosition*)positionFromPosition:(UITextPosition*)position offset:(NSInteger)offset` with offset is `-1`
- `- (UITextRange*)textRangeFromPosition:(UITextPosition*)fromPosition toPosition:(UITextPosition*)toPosition`
- `- (void)setSelectedTextRange:(UITextRange*)selectedTextRange` with new textRange
- `- (void)deleteBackward`

However, in case of Thai vowel, it skipped `- (void)setSelectedTextRange:`, so `_selectedTextRange` is going to be  empty.
Moreover, when I modified `- (UITextPosition*)positionFromPosition` to return previous codepoint, I found that it happened to call `- (void)setSelectedTextRange:` correctly.

So, I think `UIKeyboardImpl` did some magical things that made them be able to distinguish between grapheme cluster that had to be completely deleted, and partly deleted.

I think it's not going to break current behavior because it called `- (void)setSelectedTextRange:`, so `_selectedTextRange` wouldn't be empty for other cases.